### PR TITLE
[BugFix] Apply ReduceCastRule to child exprs of SetOperationRelation

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #
https://github.com/StarRocks/starrocks/issues/16947
## Problem Summary(Required) ：

RelationTransformer forgets to apply ReduceCastRule to child exprs of SetOperationRelation, so a cast expr(something like cast(varchar(-1) as varchar(2)) is propagated to BE, and VectorizedCastToString::_evaluate_string is invoked to process this cast expr and the input strings are  truncated unexpectedly.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [x] 2.4
  - [x] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by satanson** [comment]
@mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 18 / 20 (90.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Type.java|7|8|87.50%|[1001]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|8|9|88.89%|[264]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|3|3|100.00%|[]|

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 19 / 21 (90.48%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Type.java|7|8|87.50%|[1001]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|9|10|90.00%|[265]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|3|3|100.00%|[]|

File: be/src/exprs/cast_expr.cpp [hunk 1]
```diff
@@ -1335,27 +1335,15 @@ class VectorizedCastToStringExpr final : public Expr {
...
-    }
+    // cast(string as string) is trivial operation, just return the input column.
+    // This behavior is not compatible with MySQL
+    // 1. cast(string as varchar(n)) supported in SR, but not supported in MySQL
+    // 2. cast(string as char(n)) supported in both SR and MySQL, but in SR, in some queries, length
+    //    of char is neglected. in MySQL, the input string shall be truncated if its length is larger than
+    //    length of char.
+    // In SR, behaviors of both cast(string as varchar(n)) and cast(string as char(n)) keep the same: neglect
+    // of the length of char/varchar and return input column directly.
+    ColumnPtr _evaluate_string(ExprContext* context, const ColumnPtr& column) { return column; }
```
Comment (packy92) [inline be/src/exprs/cast_expr.cpp:1346]: We need add this desc for cast function on my document.

Comment (satanson) [reply be/src/exprs/cast_expr.cpp:1346]: cast(string as string) should be folded by ReduceCastRule.

**Review by Seaven — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.4

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.3

**Pull request merged**